blog/04-24-2023-effective-smart-contract-testing-with-chai/effective-smart-contract-testing-with-chai.md

```diff
@@ -0,0 +1,446 @@
+    const tx = await transaction.wait();
+    const tokenId =  tx.events[0].args[2].toNumber();
+```
+
+The `await` keyword is used to wait for the `listToken` function on the `market` contract instance to be executed. This function takes three arguments - the address of the NFT contract (`nftContractAddress`), the ID of the NFT (`1`), and the auction price (`auctionPrice`). This function call will list the NFT with the given ID for sale on the `market` contract instance at the specified auction price.
+
+The next line `const tx = await transaction.wait();` waits for the transaction to be mined on the blockchain and then assigns the transaction object to a variable named tx.
+
+The `tokenId` variable is then assigned the ID of the newly listed NFT by accessing the `args` object of the first event emitted by the `wait` function call. The `toNumber()` function is used to convert the ID from a BigNumber object to a regular number.
+
+Following this, this code is used to simulate two different buyers purchasing the NFT that was listed for sale in the previous code block.
+
+```js
+    await market.connect(acc1).buyToken(0, { value: auctionPrice });
+    const t = await myNFT.connect(acc1).resaleApproval(tokenId);
+    await market.connect(acc2).buyToken(0, { value: auctionPrice });
+```
+
+The first line uses the `await` keyword to wait for the `buyToken` function on the `market` contract instance to be executed. This function call is executed by `acc1`, one of the accounts created earlier. The `buyToken` function takes two arguments - the ID of the NFT being purchased (0) and an object with the value key set to the `auctionPrice` that was previously set for the NFT. This transaction will transfer the NFT ownership from the original owner to acc1.
+
+The next line, `const t = await myNFT.connect(acc1).resaleApproval(tokenId);` approves the resale of the NFT by `acc1`. This step is necessary because the NFT now belongs to `acc1` and can only be resold by their approval. The `resaleApproval` function on the `myNFT` contract instance is used to give this approval, with the `tokenId` being passed as the only argument.
+
+The third line of code executes the purchase of the same NFT by a different account, `acc2`. The `buyToken` function is called again on the `market` contract instance, this time with `acc2` as the buyer. The NFT is transferred to `acc2` and the purchase is complete.
+
+After this, you would want to print the URIs of all active listings on our NFT marketplace to the console, allowing users to browse and purchase NFTs.
+
+```js
+    const listingLength = await market.getListingLength();
+
+    for (let index = 0; index < listingLength; index++) {
+      const listing = await market.getListing(index);
+      const [seller, owner, token, tokenId, price, sold] = listing;
+      // console.log(seller, owner, token, tokenId);
+      const tokenUri = await myNFT.tokenURI(tokenId);
+      console.log(tokenUri);
+        }
+    });
+
+```
```

Comment:
```suggestion
```js
// Previous code
const listingLength = await market.getListingLength();

for(let index = 0; index < listingLength; index++) {
    const listing = await market.getListing(index);
    const [seller, owner, token, tokenId, price, sold] = listing;
    const tokenUri = await myNFT.tokenURI(tokenId);
    console.log(tokenURI);
}
```
```